### PR TITLE
update to Legolas v0.5

### LINKED ISSUE #20: Upgrade to Legolas 0.5
Legolas 0.5 changes how schemas are defined and used.  There's a single schema defined here, `ModelRow`, which needs to be upgraded to use `@schema`/`@version` instead of `@row`.

### PR DESCRIPTION
**Pull request by ericphanson**
Breakingness is debatable, re- https://github.com/beacon-biosignals/LegolasFlux.jl/issues/20

I think these are all the changes needed? Not 100% sure though. Tests pass at least.

closes #20

### CODE REVIEW

**Comment by ericphanson** [comment]
Maybe we should test that an old table can still be loaded. Could save out the DigitsModel from the example on main, check it in as a test, and then make sure we can load it here?

File: src/LegolasFlux.jl [hunk 1]
```diff
@@ -79,12 +78,19 @@ ArrowTypes.arrowname(::Type{<:Weights}) = WEIGHTS_ARROW_NAME
...
+##### `ModelV1`
 #####
 
-const ModelRow = Legolas.@row("legolas-flux.model@1",
-                              weights::Union{Missing,Weights} = ismissing(weights) ? missing : Weights(weights),
-                              architecture_version::Union{Missing,Int})
+@schema "legolas-flux.model" Model
+
+@version ModelV1 begin
+    weights::(<:Union{Missing,Weights}) = ismissing(weights) ? missing : Weights(weights)
```
Comment (ericphanson) [inline src/LegolasFlux.jl:87]: this is to parametrize the structs type on the weights, which I think is probably worthwhile (and Legolas pre-v0.5 did)

**Comment by kleinschmidt** [comment]
> Maybe we should test that an old table can still be loaded. Could save out the DigitsModel from the example on main, check it in as a test, and then make sure we can load it here?

was goign to suggest exactly this.

I think it's also good to use the digitsmodel as a test for extensibility

File: README.md [hunk 2]
```diff
@@ -75,7 +75,7 @@ const DigitsRow = Legolas.@row("digits.model@1" > "legolas-flux.model@1",
                          commit_sha::Union{Missing, String})
```
Comment (kleinschmidt) [inline README.md:75]: need to update this example too

File: README.md [hunk 3]
```diff
@@ -29,8 +29,8 @@ my_model = make_my_model()
 using LegolasFlux
 
 # We can save whatever other columns we'd like to as well as the `weights`.
-model_row = ModelRow(; weights = fetch_weights(cpu(my_model)),
-                     architecture_version=1, loss=0.5)
+model_row = LegolasFlux.ModelV1(; weights = fetch_weights(cpu(my_model)),
+                                architecture_version=1, loss=0.5)
```
Comment (kleinschmidt) [inline README.md:33]: huh, did we used to export ModelRow?  Shoudl we?

Comment (kleinschmidt) [reply README.md:33]: also, this will now silently drop the loss right?  sicne it's not included in the schema

**Comment by kleinschmidt** [comment]
> I think it's also good to use the digitsmodel as a test for extensibility

OIC this is included in teh tests, very good

File: README.md [hunk 4]
```diff
@@ -50,13 +50,13 @@ to keep track of for which architectures the weights are valid for.
 
 See [examples/digits.jl](examples/digits.jl) for a larger example.
 
-## `LegolasFlux.ModelRow`
+## `LegolasFlux.ModelV1`
```
Comment (kleinschmidt) [inline README.md:53]: ah yeah, even though this isn't exported I think it's probably safer to tag this as breaking.

**Review by kleinschmidt — COMMENTED** [review]
I think this generally looks good; just needs a few more updates in readme and maybe elsewhere (on exensibility)

I'd also add tests to make sure we're explicit about the "row record construcotr drops non-schema fields" behavior change (it's referenced in the readme at least)

File: README.md (hunk 3 above)
Comment (ericphanson) [reply README.md:33]: Ohhh right, it will drop the loss. We did not export it; that it wasn't qualified here was a bug

File: README.md (hunk 4 above)
Comment (ericphanson) [reply README.md:53]: We could continue to keep the alias, but I am afraid of having no backport version #s

File: examples/digits.jl [hunk 5]
```diff
@@ -56,19 +59,34 @@ end
...
+@schema "digits.model" DigitsRow
+
+compat_config(config::DigitsConfigV1) = config
+function compat_config(config::NamedTuple)
+    # This is how these deserialize from Legolas v0.4
+    if haskey(config, 1) && config[1] == "digits-config" && haskey(config, 2) &&
+       config[2] == 1
+        return DigitsConfigV1(config[3])
+    else
+        # We have some other NamedTuple, possibly from an earlier Legolas version. Trying constructing ` DigitsConfigV1`.
```
Comment (kleinschmidt) [inline examples/digits.jl:71]:
```suggestion
        # We have some other NamedTuple, possibly from an earlier Legolas version. Trying constructing `DigitsConfigV1`.
```

File: examples/digits.jl [hunk 6]
```diff
@@ -56,19 +59,34 @@ end
 
 # Here, we define a schema extension of the `legolas-flux.model` schema.
 # We add our `DigitsConfig` object, as well as the epoch and accuracy.
-const DigitsRow = Legolas.@row("digits.model@1" > "legolas-flux.model@1",
-                               config::DigitsConfig = DigitsConfig(config),
-                               epoch::Union{Missing,Int},
-                               accuracy::Union{Missing,Float32})
+@schema "digits.model" DigitsRow
+
+compat_config(config::DigitsConfigV1) = config
```
Comment (kleinschmidt) [inline examples/digits.jl:64]:
```suggestion
# construct a `DigitsConfigV1` with support for older Legolas serialization formats
# to support backwards compatibility
compat_config(config::DigitsConfigV1) = config
```

File: examples/digits.jl [hunk 7]
```diff
@@ -56,19 +59,34 @@ end
...
+        # We have some other NamedTuple, possibly from an earlier Legolas version. Trying constructing ` DigitsConfigV1`.
+        return DigitsConfigV1(config)
+    end
+end
+
+@version DigitsRowV1 > LegolasFlux.ModelV1 begin
+    config::Union{<:NamedTuple,DigitsConfigV1} = compat_config(config)
+    epoch::Union{Missing,Int}
+    accuracy::Union{Missing,Float32}
+end
```
Comment (kleinschmidt) [inline examples/digits.jl:80]: should we also parametrize the row type on the weights?  IIRC that requires re-declaring the field here

```suggestion
@version DigitsRowV1 > LegolasFlux.ModelV1 begin
    # parametrize the row type on the weights type (copied from ModelV1)
    weights::(<:Union{Missing,Weights}) = ismissing(weights) ? missing : Weights(weights)
    config::Union{<:NamedTuple,DigitsConfigV1} = compat_config(config)
    epoch::Union{Missing,Int}
    accuracy::Union{Missing,Float32}
end
```

File: src/LegolasFlux.jl (hunk 1 above)
Comment (kleinschmidt) [reply src/LegolasFlux.jl:87]: let's leave a comment to that effect with a link the legolas docs?



```suggestion
    # expose the concrete type of `weights` as a type parameter of `ModelV1`
    # see https://github.com/beacon-biosignals/Legolas.jl/blob/main/examples/tour.jl#L291
    weights::(<:Union{Missing,Weights}) = ismissing(weights) ? missing : Weights(weights)
```

**Review by kleinschmidt — APPROVED** [review]
looks good.  I'll approve and leave it up to you to decide how to handle the type paremeters :)

**Review by kleinschmidt — COMMENTED** [review]
One last thing that might be a nice QOL addition is to add something (either in a NEWS.md or in the readme) that at least points to this PR as a guide for how to upgrade existing schemas for this version, and at most provides a step-by-step guide with the gotchas we've hit so far:

- type params used to be implicit in the row types, now need to manually specified (esp. in extensions)
- constructed instances of the row types no longer include "extra" columns (make sure they exist in teh schema, probably as `Union{Missing,T}`
- make sure you have reference tests deserializing old versions from arrow if that matters to you; likely need to add some special handling for any fields that are themselves legolas rows

**Comment by ericphanson** [comment]
Should that go in some Legolas guidance, or here?

**Comment by kleinschmidt** [comment]
Hm, good question.  I think there are some specific points that are particularly relevant for upgrading schemas intended to be used like the Model extensions here, so I think it's worth calling them out in as concise a way as possible, since at the moment it's a bit extract those actionable things from teh legolas documentation/tour/various PR discussions.  I do think a _general_ legolas upgrade checklist would be good to include in legolas but not that's probably a slightly larger thing

**Comment by ericphanson** [comment]
I agree the current set of

> teh legolas documentation/tour/various PR discussions

is not conducive to quickly learning how to upgrade, but I think we should PR docs there specifically, since none of the points you raised seem LegolasFlux specific. I'll merge this, then make that PR (which will use some of this PR as an example), then make another PR here linking out to those docs.

**Pull request merged**